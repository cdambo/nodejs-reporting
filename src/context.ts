export default class Context {
  private readonly globalTags: object;

  public constructor(globalTags: object = {}) {
    this.globalTags = globalTags;
  }

  public mergeTags(tags: object = {}): object {
    return Object.assign({}, this.globalTags, tags);
  }
}
