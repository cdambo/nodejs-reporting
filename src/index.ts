import Context from "./context";

export * from "./metrics";
export { Context };
