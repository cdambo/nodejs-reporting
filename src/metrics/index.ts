export * from "./formats";
export * from "./reporters";
export * from "./integrations";
