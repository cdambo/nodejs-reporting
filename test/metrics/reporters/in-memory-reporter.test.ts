import { invoke } from "lodash";
import { InMemoryReporter } from "../../../src";

const getReporter = (): InMemoryReporter =>
  new InMemoryReporter({
    globalTags: { globalTag1: "global_one", globalTag2: "global_two" }
  });

const reportsOf = (
  method: keyof InMemoryReporter,
  ...args: unknown[]
): unknown[] => {
  const reporter = getReporter();
  invoke(reporter, method, ...args);
  return reporter.reports;
};

describe("InMemoryReporter", (): void => {
  describe("Constructor", (): void => {
    it("Creates an InMemoryReporter", (): void => {
      const reporter = new InMemoryReporter();
      expect(reporter).toMatchSnapshot();
    });

    it("Creates an InMemoryReporter with global tags", (): void => {
      const reporter = new InMemoryReporter({
        globalTags: { globalTag1: "global_one", globalTag2: "global_two" }
      });
      expect(reporter).toMatchSnapshot();
    });

    it("Creates an InMemoryReporter with a format", (): void => {
      const format = { format: (): object => ({ format: "Format" }) };
      const reporter = new InMemoryReporter({ format });
      expect(reporter).toMatchSnapshot();
    });
  });

  describe("timing", (): void => {
    it("Reports a timing metric", (): void => {
      expect(reportsOf("timing", "timing_test", 20)).toMatchSnapshot();
    });

    it("Reports a timing metric with a sample rate", (): void => {
      expect(reportsOf("timing", "timing_test", 20, 0.5)).toMatchSnapshot();
    });

    it("Reports a timing metric with tags", (): void => {
      expect(
        reportsOf("timing", "timing_test", 20, undefined, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("increment", (): void => {
    it("Reports a counter metric", (): void => {
      expect(reportsOf("increment", "increment_test")).toMatchSnapshot();
    });

    it("Reports a counter metric with a sample rate", (): void => {
      expect(reportsOf("increment", "increment_test", 0.5)).toMatchSnapshot();
    });

    it("Reports a counter metric with tags", (): void => {
      expect(
        reportsOf("increment", "increment_test", undefined, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("incrementBy", (): void => {
    it("Reports a counter metric", (): void => {
      expect(
        reportsOf("incrementBy", "increment_by_test", 10)
      ).toMatchSnapshot();
    });

    it("Reports a counter metric with tags", (): void => {
      expect(
        reportsOf("incrementBy", "increment_by_test", 10, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("decrement", (): void => {
    it("Reports a counter metric", (): void => {
      expect(reportsOf("decrement", "decrement_test")).toMatchSnapshot();
    });

    it("Reports a counter metric with a sample rate", (): void => {
      expect(reportsOf("decrement", "decrement_test", 0.5)).toMatchSnapshot();
    });

    it("Reports a counter metric with tags", (): void => {
      expect(
        reportsOf("decrement", "decrement_test", undefined, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("decrementBy", (): void => {
    it("Reports a counter metric", (): void => {
      expect(
        reportsOf("decrementBy", "decrement_by_test", 10)
      ).toMatchSnapshot();
    });

    it("Reports a counter metric with tags", (): void => {
      expect(
        reportsOf("decrementBy", "decrement_by_test", 10, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("gauge", (): void => {
    it("Reports a gauge metric", (): void => {
      expect(reportsOf("gauge", "gauge_test", 10)).toMatchSnapshot();
    });

    it("Reports a gauge metric with a sample rate", (): void => {
      expect(reportsOf("gauge", "gauge_test", 10, 0.5)).toMatchSnapshot();
    });

    it("Reports a gauge metric with tags", (): void => {
      expect(
        reportsOf("gauge", "gauge_test", 10, undefined, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("histogram", (): void => {
    it("Reports a histogram metric", (): void => {
      expect(reportsOf("histogram", "histogram_test", 20)).toMatchSnapshot();
    });

    it("Reports a histogram metric with a sample rate", (): void => {
      expect(
        reportsOf("histogram", "histogram_test", 20, 0.5)
      ).toMatchSnapshot();
    });

    it("Reports a histogram metric with tags", (): void => {
      expect(
        reportsOf("histogram", "histogram_test", 20, undefined, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("flush", (): void => {
    it("Flushes all reports from memory", (): void => {
      const reporter = getReporter();
      reporter.timing("timing_test", 20, 0.5, { tag1: "one", tag2: "two" });
      reporter.increment("increment_test", 0.5, { tag1: "one", tag2: "two" });
      reporter.decrement("decrement_test", 0.5, { tag1: "one", tag2: "two" });
      reporter.gauge("gauge_test", 10, 0.5, { tag1: "one", tag2: "two" });
      reporter.histogram("histogram_test", 20, 0.5, {
        tag1: "one",
        tag2: "two"
      });
      reporter.flush();
      expect(reporter.reports).toMatchSnapshot();
    });
  });

  describe("reports", (): void => {
    it("Saves all reports in memory", (): void => {
      const reporter = getReporter();
      reporter.timing("timing_test", 20, 0.5, { tag1: "one", tag2: "two" });
      reporter.increment("increment_test", 0.5, { tag1: "one", tag2: "two" });
      reporter.decrement("decrement_test", 0.5, { tag1: "one", tag2: "two" });
      reporter.gauge("gauge_test", 10, 0.5, { tag1: "one", tag2: "two" });
      reporter.histogram("histogram_test", 20, 0.5, {
        tag1: "one",
        tag2: "two"
      });
      expect(reporter.reports).toMatchSnapshot();
    });
  });
});
