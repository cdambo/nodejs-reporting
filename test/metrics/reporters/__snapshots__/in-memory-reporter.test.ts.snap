// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InMemoryReporter Constructor Creates an InMemoryReporter 1`] = `
InMemoryReporter {
  "context": Context {
    "globalTags": Object {},
  },
  "format": Object {
    "format": [Function],
  },
  "reports": Array [],
}
`;

exports[`InMemoryReporter Constructor Creates an InMemoryReporter with a format 1`] = `
InMemoryReporter {
  "context": Context {
    "globalTags": Object {},
  },
  "format": Object {
    "format": [Function],
  },
  "reports": Array [],
}
`;

exports[`InMemoryReporter Constructor Creates an InMemoryReporter with global tags 1`] = `
InMemoryReporter {
  "context": Context {
    "globalTags": Object {
      "globalTag1": "global_one",
      "globalTag2": "global_two",
    },
  },
  "format": Object {
    "format": [Function],
  },
  "reports": Array [],
}
`;

exports[`InMemoryReporter decrement Reports a counter metric 1`] = `
Array [
  Object {
    "decrement": Object {
      "sampleRate": 1,
      "stat": "decrement_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
    },
  },
]
`;

exports[`InMemoryReporter decrement Reports a counter metric with a sample rate 1`] = `
Array [
  Object {
    "decrement": Object {
      "sampleRate": 0.5,
      "stat": "decrement_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
    },
  },
]
`;

exports[`InMemoryReporter decrement Reports a counter metric with tags 1`] = `
Array [
  Object {
    "decrement": Object {
      "sampleRate": 1,
      "stat": "decrement_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
    },
  },
]
`;

exports[`InMemoryReporter decrementBy Reports a counter metric 1`] = `
Array [
  Object {
    "decrement": Object {
      "stat": "decrement_by_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
      "value": 10,
    },
  },
]
`;

exports[`InMemoryReporter decrementBy Reports a counter metric with tags 1`] = `
Array [
  Object {
    "decrement": Object {
      "stat": "decrement_by_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
      "value": 10,
    },
  },
]
`;

exports[`InMemoryReporter flush Flushes all reports from memory 1`] = `Array []`;

exports[`InMemoryReporter gauge Reports a gauge metric 1`] = `
Array [
  Object {
    "gauge": Object {
      "sampleRate": 1,
      "stat": "gauge_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
      "value": 10,
    },
  },
]
`;

exports[`InMemoryReporter gauge Reports a gauge metric with a sample rate 1`] = `
Array [
  Object {
    "gauge": Object {
      "sampleRate": 0.5,
      "stat": "gauge_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
      "value": 10,
    },
  },
]
`;

exports[`InMemoryReporter gauge Reports a gauge metric with tags 1`] = `
Array [
  Object {
    "gauge": Object {
      "sampleRate": 1,
      "stat": "gauge_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
      "value": 10,
    },
  },
]
`;

exports[`InMemoryReporter histogram Reports a histogram metric 1`] = `
Array [
  Object {
    "histogram": Object {
      "sampleRate": 1,
      "stat": "histogram_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
      "time": 20,
    },
  },
]
`;

exports[`InMemoryReporter histogram Reports a histogram metric with a sample rate 1`] = `
Array [
  Object {
    "histogram": Object {
      "sampleRate": 0.5,
      "stat": "histogram_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
      "time": 20,
    },
  },
]
`;

exports[`InMemoryReporter histogram Reports a histogram metric with tags 1`] = `
Array [
  Object {
    "histogram": Object {
      "sampleRate": 1,
      "stat": "histogram_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
      "time": 20,
    },
  },
]
`;

exports[`InMemoryReporter increment Reports a counter metric 1`] = `
Array [
  Object {
    "increment": Object {
      "sampleRate": 1,
      "stat": "increment_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
    },
  },
]
`;

exports[`InMemoryReporter increment Reports a counter metric with a sample rate 1`] = `
Array [
  Object {
    "increment": Object {
      "sampleRate": 0.5,
      "stat": "increment_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
    },
  },
]
`;

exports[`InMemoryReporter increment Reports a counter metric with tags 1`] = `
Array [
  Object {
    "increment": Object {
      "sampleRate": 1,
      "stat": "increment_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
    },
  },
]
`;

exports[`InMemoryReporter incrementBy Reports a counter metric 1`] = `
Array [
  Object {
    "increment": Object {
      "stat": "increment_by_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
      "value": 10,
    },
  },
]
`;

exports[`InMemoryReporter incrementBy Reports a counter metric with tags 1`] = `
Array [
  Object {
    "increment": Object {
      "stat": "increment_by_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
      "value": 10,
    },
  },
]
`;

exports[`InMemoryReporter reports Saves all reports in memory 1`] = `
Array [
  Object {
    "time": Object {
      "sampleRate": 0.5,
      "stat": "timing_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
      "time": 20,
    },
  },
  Object {
    "increment": Object {
      "sampleRate": 0.5,
      "stat": "increment_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
    },
  },
  Object {
    "decrement": Object {
      "sampleRate": 0.5,
      "stat": "decrement_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
    },
  },
  Object {
    "gauge": Object {
      "sampleRate": 0.5,
      "stat": "gauge_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
      "value": 10,
    },
  },
  Object {
    "histogram": Object {
      "sampleRate": 0.5,
      "stat": "histogram_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
      "time": 20,
    },
  },
]
`;

exports[`InMemoryReporter timing Reports a timing metric 1`] = `
Array [
  Object {
    "time": Object {
      "sampleRate": 1,
      "stat": "timing_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
      "time": 20,
    },
  },
]
`;

exports[`InMemoryReporter timing Reports a timing metric with a sample rate 1`] = `
Array [
  Object {
    "time": Object {
      "sampleRate": 0.5,
      "stat": "timing_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
      },
      "time": 20,
    },
  },
]
`;

exports[`InMemoryReporter timing Reports a timing metric with tags 1`] = `
Array [
  Object {
    "time": Object {
      "sampleRate": 1,
      "stat": "timing_test",
      "tags": Object {
        "global_tag_1": "global_one",
        "global_tag_2": "global_two",
        "tag_1": "one",
        "tag_2": "two",
      },
      "time": 20,
    },
  },
]
`;
